hibernate-core/src/main/java/org/hibernate/tool/schema/internal/ColumnDefinitions.java

```diff
@@ -21,13 +21,29 @@
 
 import java.util.List;
 import java.util.Locale;
+import java.util.StringTokenizer;
 
 import static org.hibernate.type.SqlTypes.isNumericOrDecimal;
 import static org.hibernate.type.SqlTypes.isStringType;
 
+/**
+ * Util class for comparing required vs actual column definition
+ *
+ * @author Gavin King
+ * @author Cedomir Igaly
+ * @author Steve Ebersole
+ * @author Andrea Boriero
+ * @author Yanming Zhou
+ */
 class ColumnDefinitions {
 
 	static boolean hasMatchingType(Column column, ColumnInformation columnInformation, Metadata metadata, Dialect dialect) {
+		if (column.getTypeName() != null) {
+			// @Column(columnDefinition) present
+			String required = new StringTokenizer( normalize( column.getTypeName() ), "( " ).nextToken();
+			String actual = normalize( columnInformation.getTypeName() );
+			return actual.equals( required );
+		}
```

Comment:
I don't understand the purpose of this code, and naively it doesn't look like it should be necessary.

It should already be taken care of by the existing:

    normalize( stripArgs( column.getSqlType( metadata ) ) ).equals( normalize( columnInformation.getTypeName() ) );

Since `column.getSqlType( metadata )` returns the `sqlTypeName` if there is one, and since `stripArgs(x)` does the same thing as `new StringTokenizer(x , "( " ).nextToken()`.

Reply:
> It should already be taken care of by the existing:
> 
> ```java
> normalize( stripArgs( column.getSqlType( metadata ) ) ).equals( normalize( columnInformation.getTypeName() ) );
> ```

1. It should return `false` immediately instead of fallback if `@Column(columnDefinition)` present.
2. `stripArgs()` doesn't strip ` not null`.

Reply:
Well, if that's the case then better fixes would be:

1. Add `else if (column.getSqlTypeName()!=null) { return false; }` which is much simpler.
2. Change the `stripArgs()` to handle the additional cases you're concerned about.

But the thing is I just don't see how this can be made robust. Your implementation reduces `character varying (10) not null` to just `character`, which is a different type. So that's no good. Now, sure, you could look for `not null` explicitly and remove it, but there's many more complicated things apart from `not null` that this freeform `columnDefinition` could contain.

I mean, if you can think of something, I'm all ears, but just "take the first token" definitely isn't good enough.

Reply:
I added dedicated method `extractTypeName(String typeExpression)` to handle this.